### PR TITLE
Wire tool improvements

### PR DESCRIPTION
**Pull request by MulleDK19**
- Improved wire rendering (Wires retain their width no matter how big of an angle there is between wirepoints).
- Added visualization of wires while adding them.
- Added the ability to remove wirepoints while wiring, without having to redo the entire wire.
- Added option to stick wires to surfaces.

https://youtu.be/btE3df5xJlk

### CODE REVIEW

File: lua/weapons/gmod_tool/stools/wire_adv.lua [hunk 1]
```diff
@@ -17,13 +17,15 @@ if CLIENT then
...
 	language.Add( "WireTool_colour", "Colour:" )
+	language.Add( "WireTool_stick", "Stick to surfaces" )
 	TOOL.Information = {
 		{ name = "left_0", stage = 0, text = "Select input (Shift: Select multiple; Alt: Select all)" },
 		{ name = "right_0", stage = 0, text = "Next" },
 		{ name = "reload_0", stage = 0, text = "Unlink (Alt: Unlink all)" },
 		{ name = "mwheel_0", stage = 0, text = "Mouse wheel: Next" },
 		{ name = "left_1", stage = 1, text = "Select entity" },
 		{ name = "right_1", stage = 1, text = "Add wirepoint" },
+		{ name = "f_0", stage = 1, text = "F: Remove wirepoint" },
```
Comment (Divran) [inline lua/weapons/gmod_tool/stools/wire_adv.lua]: wouldn't it make more sense to say "undo"?

File: lua/weapons/gmod_tool/stools/wire_adv.lua [hunk 2]
```diff
@@ -413,6 +418,62 @@ elseif CLIENT then
...
+			normal = trace.Entity:GetRight()
+		elseif dir == 3 then
+			normal = -trace.Entity:GetUp()
+		elseif dir == 4 then
+			normal = -trace.Entity:GetForward()
+		elseif dir == 5 then
+			normal = -trace.Entity:GetRight()
+		else -- if 0 then
+			normal = trace.Entity:GetUp()
+		end
```
Comment (Divran) [inline lua/weapons/gmod_tool/stools/wire_adv.lua]: What is the point of the above 10-ish lines? Wouldn't this always be `normal = trace.HitNormal` at all times?

File: lua/weapons/gmod_tool/stools/wire_adv.lua (hunk 1 above)
Comment (MulleDK19) [reply lua/weapons/gmod_tool/stools/wire_adv.lua]: Yes

File: lua/weapons/gmod_tool/stools/wire_adv.lua (hunk 2 above)
Comment (MulleDK19) [reply lua/weapons/gmod_tool/stools/wire_adv.lua]: No. This function traces the surface of the parent entity no matter where you hit the wire entity. For most gates the parent is in their "down" direction, but not for all. For some it's forward, etc. This function traces all directions from the hit point on the wire entity, to find the surface of the parent, to stick wires to their surface.

File: lua/weapons/gmod_tool/stools/wire_adv.lua [hunk 3]
```diff
@@ -413,6 +418,62 @@ elseif CLIENT then
 		return self.AutoWiringTypeLookup_t[inputtype]
 	end
 
+	function TOOL:IsWireEntity(ent)
```
Comment (Divran) [inline lua/weapons/gmod_tool/stools/wire_adv.lua]: This function already exists https://github.com/wiremod/wire/blob/master/lua/wire/wireshared.lua#L451

File: lua/weapons/gmod_tool/stools/wire_adv.lua (hunk 2 above)
Comment (Divran) [reply lua/weapons/gmod_tool/stools/wire_adv.lua]: Alright, but using recursion to solve this seems a bit strange. Personally I'd use regular old iteration instead to make it clearer what the code does, but the performance impact difference will be negligible.

Comment (MulleDK19) [reply lua/weapons/gmod_tool/stools/wire_adv.lua]: Alright.

File: lua/weapons/gmod_tool/stools/wire_adv.lua (hunk 3 above)
Comment (MulleDK19) [reply lua/weapons/gmod_tool/stools/wire_adv.lua]: Cheers.

File: lua/weapons/gmod_tool/stools/wire_adv.lua (hunk 2 above)
Comment (MulleDK19) [reply lua/weapons/gmod_tool/stools/wire_adv.lua]: Decided to keep it. It'd be more messy as an iteration, since it has to restart, etc.

**Comment by MulleDK19** [comment]
> @thegrb93 this has been in PR hell for long enough. What's your reason for adding the "Needs work" label to it? Maybe we can merge it now?
> 
> I've completely forgotten about this PR so I haven't tested it, but we do have a video of it working and the code looks fine so I think we can trust the author here.

I've been busy with stuff. I haven't merged it with the latest master yet; and I think I may have some local changes too. I'll see if I can take a look at it tomorrow.

**Comment by thegrb93** [comment]
@Divran need to resolve conflicts

**Comment by MulleDK19** [comment]
Sorry for the masive delay. I've been quite busy elsewhere, but I finally got around to this.
I had a lot of trouble getting this merged right with my changes being so out of date, and I had to force a push, as I think I ended up with duplicate commits from a rebase.

Not sure if this should be pulled, or if I should backup the branch, delete it, recreate a new branch and then redo my changes on that to get some clean commits. Perhaps someone with more Git knowledge can give some advice here.

**Comment by thegrb93** [comment]
Can you explain the purpose of UpdateTraceForSurface again? I'm a bit lost.

**Comment by MulleDK19** [comment]
> Can you explain the purpose of UpdateTraceForSurface again? I'm a bit lost.

It updates the trace from whatever the tool hit, to the hit entity's parent surface. It aligns the wires to the surfaces underneath whatever you're wiring. <https://youtu.be/btE3df5xJlk?t=30>

**Comment by thegrb93** [comment]
Ok I see. Instead of the recursion and if/else, can you just put the directions in a table and loop over them? After that I think it'd be good.

**Comment by MulleDK19** [comment]
> Ok I see. Instead of the recursion and if/else, can you just put the directions in a table and loop over them? After that I think it'd be good.

Putting them in a table would calculate all normals even when only the first few are needed.
And I think it's gonna be a lot more messy with a loop. But I'll try and see.

How about the commits? I had to force push it, and I don't know if it ended up with duplicate commits.

**Comment by MulleDK19** [comment]
In either case, I've changed UpdateTraceForSurface from using recursion to using a loop, and fixed some issues with it.

**Comment by Divran** [comment]
We do have continue thanks to garry, and everyone's fine with using it as far as I know because it's a feature lua should have by default anyway.

File: lua/weapons/gmod_tool/stools/wire_adv.lua [hunk 4]
```diff
@@ -417,57 +417,74 @@ elseif CLIENT then
...
+			local normal = directions[dir + 1]
+
+			traceData.endpos = trace.HitPos + normal * -1000
+			traceData.filter = { LocalPlayer(), trace.Entity }
+			traceData.collisiongroup = LAST_SHARED_COLLISION_GROUP
+			local newTrace = util.TraceLine(traceData)
+
+			local foundParent = false
+			if newTrace.Hit and IsValid(parent) and newTrace.Entity == parent then
+				local distanceSquared = (newTrace.HitPos - trace.HitPos):LengthSqr()
```
Comment (thegrb93) [inline lua/weapons/gmod_tool/stools/wire_adv.lua]: Vector:DistToSqr would be better

**Comment by thegrb93** [comment]
Not sure the exact logic you want, but this structure should make things easier to follow.

```lua
function TOOL:UpdateTraceForSurface(trace, parent)
	if self:GetClientNumber("stick") == 0 then return end
	if not WireLib.HasPorts(trace.Entity) then return end
	
	local entityUp = trace.Entity:GetUp() * 1000
	local entityForward = trace.Entity:GetForward() * 1000
	local entityRight = trace.Entity:GetRight() * 1000
	local directions =
	{
		entityUp,
		entityForward,
		entityRight,
		-entityUp,
		-entityForward,
		-entityRight
	}

	local function findHit(parent)
		local traceData = util.GetPlayerTrace(LocalPlayer())
		traceData.start = trace.HitPos
		traceData.filter = { LocalPlayer(), trace.Entity }
		traceData.collisiongroup = LAST_SHARED_COLLISION_GROUP
		
		local foundParent = false
		local hitParentPos
		local hitParentNormal
		local closestDistanceSquared = 99999999
		for dir=1, 6 do

			traceData.endpos = trace.HitPos + directions[dir]
			local newTrace = util.TraceLine(traceData)

			if newTrace.Entity == parent then
				local distanceSquared = newTrace.HitPos:DistToSqr(trace.HitPos)
				if distanceSquared < closestDistanceSquared then
					closestDistanceSquared = distanceSquared
					hitParentPos = newTrace.HitPos
					hitParentNormal = newTrace.HitNormal
					foundParent = true
				end
			end
		end
		return foundParent, hitParentPos, hitParentNormal
	end
	
	local foundParent, hitParentPos, hitParentNormal

	if IsValid(parent) then
		foundParent, hitParentPos, hitParentNormal = findHit(parent)
	end

	if not foundParent then
		-- Didn't find the parent in any direction, assume whichever entity can be traced behind the entity is the "parent".
		-- This can happen if eg. the component is not directly on the parent.
		local traceData = util.GetPlayerTrace(LocalPlayer())
		traceData.filter = { LocalPlayer(), trace.Entity } -- Ignore player and the entity we're trying to wire to.
		traceData.collisiongroup = LAST_SHARED_COLLISION_GROUP
		foundParent, hitParentPos, hitParentNormal = findHit(util.TraceLine(traceData).Entity)
	end

	if foundParent then
		trace.HitPos = hitParentPos
		trace.Normal = hitParentNormal
	end
end
```

**Comment by MulleDK19** [comment]
I've rewritten UpdateTraceForSurface keeping some of the suggestions in mind, as well as fixing some issues the old function had.

I've also made other wires darkened while wiring, to make it easier to see which wire you're wiring throughout.

**Comment by thegrb93** [comment]
Three problems

Can't put a wire point on the world
The currently being placed wire's material is wrong
While wiring, all other wires go dark.

**Comment by MulleDK19** [comment]
> Can't put a wire point on the world
You've never been able to.

> The currently being placed wire's material is wrong
Hmm..?

> While wiring, all other wires go dark.
That's intentional. It helps you know which wire you're working on, instead of 50 wires crossing your wire.

**Comment by thegrb93** [comment]
For some reason the wire material is super stretched while positioning it.

**Comment by MulleDK19** [comment]
🤔I don't think it's stretched for me. But I'll take a look.

**Comment by MulleDK19** [comment]
I'm unable to reproduce materials getting stretched.

image

**Comment by thegrb93** [comment]
It's only the one currently being placed. It's because you're not incrementing the scroll between lines 1147 and 1158/1162

**Pull request merged**